include/pybind11/async.h

```diff
@@ -0,0 +1,208 @@
+        async_function(Return (Class::*f)(Arg...) const&, const Extra&... extra) {
+            initialize([f](const Class *c, Arg... args) -> Return { return (c->*f)(args...); },
+                    (Return (*)(const Class *, Arg ...)) nullptr, extra...);
+        }
+
+    protected:
+        template <typename Func, typename Return, typename... Args, typename... Extra>
+        void initialize(Func &&f, Return (*)(Args...), const Extra&... extra) {
+            // create a new lambda which spawns an async thread running the original function
+            auto proxy = [f](Args... args) -> Awaitable* {
+                auto thread_func = [f](Args... args) {
+                    auto result = f(std::forward<Args>(args) ...);
+
+                    py::gil_scoped_acquire gil;
+
+                    auto py_result = py::cast(result);
+                    return py_result;
+                };
+                auto bound_thread_func = std::bind(thread_func, std::forward<Args>(args)...);
+
+                auto future = std::async(std::launch::async, bound_thread_func);
+                auto awaitable = new Awaitable(future);
+
+                return awaitable;
+            };
+
+            // initialize using the new lambda function
+            cpp_function::initialize(
+                std::forward<decltype(proxy)>(proxy),
+                (detail::function_signature_t<decltype(proxy)> *) nullptr,
+                extra...
+                );
+        }
+
+        template <typename Func, typename... Args, typename... Extra>
+        void initialize(Func &&f, void (*)(Args...), const Extra&... extra) {
+            // create a new lambda which spawns an async thread running the original function
+            auto proxy = [f](Args... args) -> Awaitable* {
+                auto thread_func = [f](Args... args) {
+                    f(std::forward<Args>(args) ...);
+
+                    py::gil_scoped_acquire gil;
+
+                    auto py_result = py::cast(Py_None);
+                    return py_result;
+                };
+                auto bound_thread_func = std::bind(thread_func, std::forward<Args>(args)...);
+
+                auto future = std::async(std::launch::async, bound_thread_func);
+                auto awaitable = new Awaitable(future);
+
+                return awaitable;
+            };
+
+            // initialize using the new lambda function
+            cpp_function::initialize(
+                std::forward<decltype(proxy)>(proxy),
+                (detail::function_signature_t<decltype(proxy)> *) nullptr,
+                extra...
+                );
+        }
+
+};
+
+template <typename type_, typename... options>
+class class_async : public class_<type_, options...> {
+
+    using type = type_;
+    public:
+        // using parent constructor
+        using class_<type_, options...>::class_;
+
+        template <typename Func, typename... Extra>
```

Comment:
Could we add it into `class_`? That would seem nicer.

Reply:
that would be a lot shorter, yes.

I think the cleanest would be to have something similar to a call_guard, but thats out of scope for me.

Reply:
Why is it out of scope? Too much work? Or something fundamental, technically?

Reply:
I'd say out of scope for my skills and I really need some time to check pybinds internals which i don't seem to have for the near future.

I think there are three things left to address:

- use the approach from @0lru
- get rid of std::bind (i tried to use lambdas but it seems to break things//segfaults//needs more work)
- make a cleaner interface (like an attribute)

Reply:
I'm not using any of these (two) approaches defined here. Maybe it's also possible to do something
with c++-coroutines(?)... However, I'm currently using an abstract Promise class. 
This has the advantage, that I don't need to start a new thread with std::async or std::thread inside the pybind11-definition:
```cpp
template<typename T>
    class Promise
    {
    public:
        class Implementation
        {
        public:
            virtual ~Implementation() = default;
            virtual void set_value(T) = 0;
            virtual void set_exception(std::exception_ptr) = 0;
        };

        Promise(std::shared_ptr<Implementation> implementation)
            : _implementation(std::move(implementation))
        {
        }

        void set_value(T t)
        {
            _implementation->set_value(std::move(t));
        }

        void set_exception(std::exception_ptr e)
        {
            _implementation->set_exception(std::move(e));
        }

    private:
        std::shared_ptr<Implementation> _implementation;
    };

... later ... :
    
    void Foo::get_bar(Promise<int>&&) const;

```

Another approach was to define a "TaskQueue" (a threaded event-loop), where "Driven" means, that
the object is assigned to such a "TaskQueue":
```cpp
 template<typename Driven, typename Param>
    auto async(py::module& asyncio, void(Driven::* member)(Param))
    {
        return [=](Driven& object, Param param) mutable
        {
            auto promise = std::make_shared<Promise<void>>(asyncio);
            auto future = promise->get_future();
            object.task_queue()->run([object = &object, member, promise{ std::move(promise) }, param{std::move(param)}]() mutable {
                try
                {
                    (*object.*member)(std::move(param));
                    promise->set_value();
                }
                catch (...)
                {
                    promise->set_exception(std::current_exception());
                }
            });
            return future;
        };
    }
```
What I want to note/say: This topic is quite complex and I want to question if std::future is
a good match for doing interaction with asyncio :-).

Reply:
> I'd say out of scope for my skills and I really need some time to check pybinds internals which i don't seem to have for the near future.

Caveat: I'm not familiar enough with threading, futures, async to have a meaningful opinion about the high-level design or implementation details.

From a high level perspective, this still looks like a lot of work to me (realistically weeks worth of time) before it could be merged. I'd be very reluctant to consider adding `class_async` as an interface, we need full comprehensive, well-written test coverage, documentation, and agreement between people with relevant experience. This is a fairly major new feature, until it's out there for a while (a few months to a couple years), we'd need some commitment to supporting it with fixes or even bigger adjustments in case early design decisions don't pan out well for general use. Do you think you are prepared for that?

Reply:
As i said, it is hard for me to find consecutive weeks of time now - during summer season, but i am motivated to push this topic forward after that.
of course the PR in it's current state is quite far from enough and as you pointed out, adding class_async doesn't seem right. I'd like to reduce the amount of code to a minimum and get rid of all copy & paste parts (class_ > class_async).

Until then, should i close this draft-pr or do you shall we leave it as a good starting-point for anyone interested / a place for discussions/ideas?

Reply:
> Until then, should i close this draft-pr or do you shall we leave it as a good starting-point for anyone interested / a place for discussions/ideas?

Up to you, keeping this open as a draft sounds good to me.

Random / general / something to think about for the maintainers:
* a `contrib/` directory might be useful for things like this, things that take a while to mature and may need the involvement of multiple people.
* having some kind of hook in the `class_` code would seem fine to me if that's helpful (as an alternative to the `call_guard` approach, or in addition to, idk), so that you could maintain your work as an add-on without being forced to introduce `class_async`.

To the second point, that's the approach I'm [currently pursuing on the smart_holder branch](https://github.com/pybind/pybind11/blob/ee945e705891cc0a5abaf8dd8174caf7cc90fb21/include/pybind11/detail/smart_holder_sfinae_hooks_only.h). (To avoid misunderstanding, the hooks are currently NOT on master, but that's where the idea comes from.)